File: 31/31.c
#include <stdio.h>
#include <stdint.h>

#define K (200)


#define N (8)
uint64_t Coins[N] = { 1, 2, 5, 10, 20, 50, 100, 200 };

uint64_t CoinSums(int64_t amount, uint64_t startIdx)
{
    if (amount < 0)
    {
        return 0;
    }
    else if (amount == 0)
    {
        return 1;
    }
    else
    {
        uint64_t count = 0;
        for (uint64_t i = startIdx; i < N; i++)
        {
            count += CoinSums(amount - Coins[i], i);
        }
        return count;
    }
}

void Solve()
{
    printf("%lu\n", CoinSums(K, 0));
}

int main()
{
    Solve();
    return 0;
}

File: 31/31.ts
const K = 200;

// Reversed order also works.
const Coins = [200, 100, 50, 20, 10, 5, 2, 1];

const CoinSums = (amount: number, startIdx: number) => {
    if (amount < 0) {
        return 0;
    }

    if (amount === 0) {
        return 1;
    }

    let sum = 0;

    for (let i = startIdx; i < Coins.length; i++) {
        sum += CoinSums(amount - Coins[i], i);
    }

    return sum;
};

(() => {
    const combinations = CoinSums(K, 0);
    console.log(combinations);
})();

export {};
